### PR TITLE
17261: add support for management access from onprem

### PR DESCRIPTION
**Pull request by shuoliugit**
- create vpc attachment with edge_attachment -> remove, import, plan -> no diff
- change edge attachment -> remove, import, plan -> no diff
- delete edge attachment -> remove, import, plan -> no diff
- create vpc attachment without edge_attachment -> remove, import, plan -> no diff
- add edge attachment -> remove, import, plan -> no diff
- destroy
- try a non firenet domain -> error

### CODE REVIEW

File: aviatrix/resource_aviatrix_aws_tgw_vpc_attachment.go [hunk 1]
```diff
@@ -245,6 +266,38 @@ func resourceAviatrixAwsTgwVpcAttachmentUpdate(d *schema.ResourceData, meta inte
...
+
+		if oldEA != "" && newEA != "" {
+			awsTgwVpcAttachment.EdgeAttachment = ""
+
+			err := client.UpdateFirewallAttachmentAccessFromOnprem(awsTgwVpcAttachment)
+			if err != nil {
+				return fmt.Errorf("failed to disable firewall attachment access from onprem while updating: %s", err)
+			}
+
+			awsTgwVpcAttachment.EdgeAttachment = d.Get("edge_attachment").(string)
```
Comment (CyrusJavan) [inline aviatrix/resource_aviatrix_aws_tgw_vpc_attachment.go]: Don't need `d.Get` can just use `newEA`

File: goaviatrix/aws_tgw_vpc_attachment.go [hunk 2]
```diff
@@ -381,3 +390,47 @@ func (c *Client) EditTgwSpokeVpcCustomizedRouteAdvertisement(awsTgwVpcAttachment
...
+	}
+
+	type AccessFromEdge struct {
+		AccessFromEdge []string `json:"access_from_edge"`
+	}
+
+	type Resp struct {
+		Return  bool             `json:"return"`
+		Results []AccessFromEdge `json:"results"`
+		Reason  string           `json:"reason"`
```
Comment (CyrusJavan) [inline goaviatrix/aws_tgw_vpc_attachment.go]: Return and Reason are unnecessary.

Also the json struct tags are unnecessary.

File: goaviatrix/aws_tgw_vpc_attachment.go [hunk 3]
```diff
@@ -170,6 +171,14 @@ func (c *Client) GetAwsTgwVpcAttachment(awsTgwVpcAttachment *AwsTgwVpcAttachment
 		aTVA.CustomizedRouteAdvertisement = ""
 	}
 
+	firenetManagementDetails, err := c.GetFirenetManagementDetails(awsTgwVpcAttachment)
+	if err != nil {
+		return nil, errors.New("could not get firenet management details: " + err.Error())
+	}
+	if len(firenetManagementDetails) != 0 {
+		aTVA.EdgeAttachment = firenetManagementDetails[1]
```
Comment (CyrusJavan) [inline goaviatrix/aws_tgw_vpc_attachment.go:179]: Why the `1` index not `0`? Will this slice always have >=2 values in it?

File: goaviatrix/aws_tgw_vpc_attachment.go [hunk 4]
```diff
@@ -170,6 +171,14 @@ func (c *Client) GetAwsTgwVpcAttachment(awsTgwVpcAttachment *AwsTgwVpcAttachment
 		aTVA.CustomizedRouteAdvertisement = ""
 	}
 
+	firenetManagementDetails, err := c.GetFirenetManagementDetails(awsTgwVpcAttachment)
+	if err != nil {
+		return nil, errors.New("could not get firenet management details: " + err.Error())
```
Comment (CyrusJavan) [inline goaviatrix/aws_tgw_vpc_attachment.go]: Can use `fmt.Errorf` formatting errors

File: goaviatrix/aws_tgw_vpc_attachment.go [hunk 5]
```diff
@@ -381,3 +390,47 @@ func (c *Client) EditTgwSpokeVpcCustomizedRouteAdvertisement(awsTgwVpcAttachment
...
+	}
+
+	var data Resp
+
+	err := c.GetAPI(&data, params["action"], params, BasicCheck)
+	if err != nil {
+		return nil, err
+	}
+
+	return data.Results[0].AccessFromEdge, nil
```
Comment (CyrusJavan) [inline goaviatrix/aws_tgw_vpc_attachment.go:433]: Will `Results` never be empty? What if the attachment is deleted on the GUI, would this panic?

File: aviatrix/resource_aviatrix_aws_tgw_vpc_attachment.go [hunk 6]
```diff
@@ -245,6 +266,38 @@ func resourceAviatrixAwsTgwVpcAttachmentUpdate(d *schema.ResourceData, meta inte
...
+
+			err = client.UpdateFirewallAttachmentAccessFromOnprem(awsTgwVpcAttachment)
+			if err != nil {
+				return fmt.Errorf("failed to enable firewall attachment access from onprem while updating: %s", err)
+			}
+		} else {
```
Comment (CyrusJavan) [inline aviatrix/resource_aviatrix_aws_tgw_vpc_attachment.go:325]: What about the case where we are going from "" -> "someID", do we need to validate again that this is a FireNet attachment?

Or would the API return an error? If that's the case we probably don't need to validate again

File: aviatrix/resource_aviatrix_aws_tgw_vpc_attachment.go [hunk 7]
```diff
@@ -80,6 +80,14 @@ func resourceAviatrixAwsTgwVpcAttachment() *schema.Resource {
 				Default:     false,
 				Description: "Advanced option. If set to true, it disables automatic route propagation of this VPC to other VPCs within the same security domain.",
 			},
+			"edge_attachment": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "",
```
Comment (CyrusJavan) [inline aviatrix/resource_aviatrix_aws_tgw_vpc_attachment.go]: Is this `Default` necessary?

File: aviatrix/resource_aviatrix_aws_tgw_vpc_attachment.go (hunk 6 above)
Comment (shuoliugit) [reply aviatrix/resource_aviatrix_aws_tgw_vpc_attachment.go:325]: API will return an error, but we can add a validation here.

File: goaviatrix/aws_tgw_vpc_attachment.go (hunk 5 above)
Comment (shuoliugit) [reply goaviatrix/aws_tgw_vpc_attachment.go:433]: I thought about this again. If the attachment is removed manually, then an error will be returned.

File: goaviatrix/aws_tgw_vpc_attachment.go (hunk 3 above)
Comment (shuoliugit) [reply goaviatrix/aws_tgw_vpc_attachment.go:179]: `0` is the security domain name, `1` is the attachment name.

File: aviatrix/resource_aviatrix_aws_tgw_vpc_attachment.go [hunk 8]
```diff
@@ -245,8 +283,57 @@ func resourceAviatrixAwsTgwVpcAttachmentUpdate(d *schema.ResourceData, meta inte
...
+			SecurityDomainName: d.Get("security_domain_name").(string),
+			EdgeAttachment:     d.Get("edge_attachment").(string),
+		}
+
+		isFirewallSecurityDomain, err := client.IsFirewallSecurityDomain(awsTgwVpcAttachment.TgwName, awsTgwVpcAttachment.SecurityDomainName)
+		if err != nil {
+			if err == goaviatrix.ErrNotFound {
+				return fmt.Errorf("could not find Security Domain: " + awsTgwVpcAttachment.SecurityDomainName)
+			}
+			return fmt.Errorf(("could not find Security Domain due to: ") + err.Error())
```
Comment (CyrusJavan) [inline aviatrix/resource_aviatrix_aws_tgw_vpc_attachment.go]: We should be consistent in our `fmt.Errorf` usage. This should just be:
```
return fmt.Errorf("could not find Security Domain due to: %v", err)
```

**Review by CyrusJavan — CHANGES_REQUESTED** [review]

**Review by CyrusJavan — APPROVED** [review]

**Review by zhiqiangjin — APPROVED** [review]

**Pull request merged**